SteamApp

camodesigns opened the pull request:
All the Header FIles
Category.h
CategoryContainer.h
Steam.h
VideoGames.h
All the cpp Files
Category.cpp
CategoryContainer.cpp
Main.cpp
Steam.cpp
VideoGames.cpp
MenuCommands Implemented
Input Validations implemented
Functions an methods created and implemented

camodesigns committed:
Steam Commit Comple

All the Header FIles
Category.h
CategoryContainer.h
Steam.h
VideoGames.h
All the cpp Files
Category.cpp
CategoryContainer.cpp
Main.cpp
Steam.cpp
VideoGames.cpp
MenuCommands Implemented
Input Validations implemented
Functions an methods created and implemented

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.h:
```diff
@@ -0,0 +1,33 @@
+	public:
+		FVideoGame();
+		FVideoGame(const std::string NewGameName, const std::string NewStudioName, const int NewDay, const int NewMonth, const int NewYear);
+		std::string GetGameName()const;
```
Remember to add some spaces to group, in this case, functions that have something in common such as grouping up certain getters for example.

JDAlzate replied:
**This is very important**

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.h:
```diff
@@ -0,0 +1,33 @@
+	private:
+		std::string GameName;
+		std::string	StudioName;
+		int DayCreation;
```
I think maybe all these 3 variables could be named creation___. For instance, CreationDay instead of DayCreation

JDAlzate replied:
I completely agree, but also it's not about the creation date but rather the release date.

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.h:
```diff
@@ -0,0 +1,33 @@
+		int SetDayCreation(const int NewDayCreation);
+		int SetMonthCreation(const int NewMonthCration);
+		int SetYearCreation(const int NewYearCreation);
+		std::string GetFormatedDate()const;
```
Formatted with a double t

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.h:
```diff
@@ -0,0 +1,33 @@
+		
+};
+
+int main()
```
Why is there a main function here??

DvdHalliday replied:
I don't think this even works since the Main.cpp is empty, and this function here is completely blank.

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.cpp:
```diff
@@ -0,0 +1,75 @@
+}
+std::string FVideoGame::GetFormatedDate()const
+{
+	std::string FormatedDate = "";
```
No need to create the variable before to another value

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.cpp:
```diff
@@ -0,0 +1,75 @@
+}
+bool FVideoGame::GameIsValid()const
+{
+	if (DayCreation <= 0 || DayCreation > 30)return false;
```
According to the standards, the content of the if statement should go in a { } body of its own

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.cpp:
```diff
@@ -0,0 +1,75 @@
+bool FVideoGame::GameIsValid()const
+{
+	if (DayCreation <= 0 || DayCreation > 30)return false;
+	else if (MonthCreation <= 0 || MonthCreation > 12)return false;
```
Also, this would deny games published on a 31st of the month, wouldn't it?

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Category.h:
```diff
@@ -0,0 +1,22 @@
+class FCategory
+{ 
+private:
+	std::string CategoryTitle;
```
I'm not sure but maybe these names shouldn't be named including the class name. If you're already in a class and you ask for its name, it's clear you won't need to know the variable name is CategoryTitle, Title is good enough.

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Category.h:
```diff
@@ -0,0 +1,22 @@
+	FCategory();
+	std::string GetCategoryTitle() const;
+	std::string SetCategoryTitle(const std::string NewCategoryTitle);
+	bool AddGame(FVideoGame VideoGame);
```
I believe this parameter could be const too

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Category.h:
```diff
@@ -0,0 +1,22 @@
+	std::string GetCategoryTitle() const;
+	std::string SetCategoryTitle(const std::string NewCategoryTitle);
+	bool AddGame(FVideoGame VideoGame);
+	FVideoGame GetGames(int GameIndex);
```
const parameter and function

JDAlzate replied:
Why is it called `GetGames` if you're just returning one game?

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Category.h:
```diff
@@ -0,0 +1,22 @@
+	std::string SetCategoryTitle(const std::string NewCategoryTitle);
+	bool AddGame(FVideoGame VideoGame);
+	FVideoGame GetGames(int GameIndex);
+	void DisplayGamesCount()const;
```
space between the () and const

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Category.cpp:
```diff
@@ -0,0 +1,49 @@
+{
+	std::cout << GamesCount << std::endl;
+}
+FVideoGame FCategory::GetGames(int GameIndex)
```
This function could be const, i didn't correct you on the .h file 'cause I wasn't sure what it did, but it definetely can be.

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Category.cpp:
```diff
@@ -0,0 +1,49 @@
+	return Games[GameIndex];
+}
+
+void   FCategory::DisplayGames()
```
same as the previous method, it should be const.

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.h:
```diff
@@ -0,0 +1,18 @@
+		const static int MaxCategoriesCount = 5;
+		FCategory Categories[MaxCategoriesCount];
+		int CategoriesCount;
+		bool IsValidIndex(int Index);
```
Const method and parameter

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.h:
```diff
@@ -0,0 +1,18 @@
+		int CategoriesCount;
+		bool IsValidIndex(int Index);
+	public:
+		void AddGameToCategory(int index,FVideoGame);
```
I think there should be a space after the , in the parameters, also why is there no name defined for the videogame parameter?

DvdHalliday replied:
the parameters could be const too

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.h:
```diff
@@ -0,0 +1,18 @@
+	public:
+		void AddGameToCategory(int index,FVideoGame);
+		FCategoryContainer();
+		void AddCategory(FCategory NewCategory);
```
Const parameter

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.h:
```diff
@@ -0,0 +1,18 @@
+		void AddGameToCategory(int index,FVideoGame);
+		FCategoryContainer();
+		void AddCategory(FCategory NewCategory);
+		void EliminateCategory(int EliminationIndex);
```
Const parameter

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.h:
```diff
@@ -0,0 +1,18 @@
+		FCategoryContainer();
+		void AddCategory(FCategory NewCategory);
+		void EliminateCategory(int EliminationIndex);
+		int GetCategoryCount()const;
```
space before const

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.h:
```diff
@@ -0,0 +1,18 @@
+		void AddCategory(FCategory NewCategory);
+		void EliminateCategory(int EliminationIndex);
+		int GetCategoryCount()const;
+		FCategory GetCategory(int Index)const;
```
space before const

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.cpp:
```diff
@@ -0,0 +1,43 @@
+FCategoryContainer::FCategoryContainer()
+{
+	CategoriesCount = 0;
+}
```
Perhaps you should add the comment separators between methods to help with readability, as well as adding white spacess!!

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.cpp:
```diff
@@ -0,0 +1,43 @@
+void FCategoryContainer::AddCategory(FCategory NewCategory)
+{
+	Categories[CategoriesCount] = NewCategory;
+	CategoriesCount += 1;
```
the ++ operator is always an option too.

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.cpp:
```diff
@@ -0,0 +1,43 @@
+}
+bool FCategoryContainer::IsValidIndex(int Index)
+{
+	if (!std::cin.good() || Index > 4 || Index < 0)return false;
```
if statement should go in a { } body

JDAlzate replied:
An index has nothing to do with the std::cin state. An index is just an index, so to check for its validity, you should be checking if it is between 0 and 4 and that's it.

After that, this code could be collapsed into a single line - that if-else is redundant.

You could just do:
```
return Index >= 0 && Index <= 4;
```

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.cpp:
```diff
@@ -0,0 +1,43 @@
+}
+void FCategoryContainer::EliminateCategory(int EliminationIndex)
+{
+	if (FCategoryContainer::IsValidIndex(EliminationIndex)) 
```
When you're defining a method of the FCategoryContainer class and you call a method of that class, there is no need to refer to that class again, however it is not wrong, just a tip in case you didn't know

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.cpp:
```diff
@@ -0,0 +1,43 @@
+{
+	if (FCategoryContainer::IsValidIndex(EliminationIndex)) 
+	{
+		for (int IndexPosition = 0;IndexPosition < MaxCategoriesCount;IndexPosition++) 
```
space after the semicolons

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.h:
```diff
@@ -0,0 +1,20 @@
+{
+private:
+	FCategoryContainer CategoryContainer;
+	enum class MenuCommand {MainMenu, AddGames, Categories ,VisualiceGames, Exit};
```
Enums should be name with the E prefix

DvdHalliday replied:
Visualize goes with a z. Also, that word might not be the right one for this, Visualize, according to cambridge dictionary means To form a picture of someone or something in your mind, in order to remember it/him/her. So perhaps you could changed it to ViewGames or DisplayGames

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+#include "Steam.h"
+
+void FSteam::DebugApp() 
```
You shouldn't have these empty functions lying around your main class

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+		case FSteam::MenuCommand::Categories:
+			OpenCategoriesMenu();
+			break;
+		case FSteam::MenuCommand::Exit:
```
is this cased ever reached?

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+	std::cout << "2.Browse Category" << std::endl;
+	std::cout << "3.Visualice Games" << std::endl;
+	std::cout << "Exit" << std::endl;
+	int OptionChoice;
```
yup, add spaces, reading this like a block is tough.

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+	std::cin.clear();
+	std::cin.ignore(1000, '\n');
+	system("CLS");
+	std::cout << "Adding a VideoGame �Fantastic! Please Follow the process " << std::endl;
```
I think the "follow the process" part could be removed

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+		std::cout << "Day: " << std::endl;
+		int EntryDay;
+		std::cin >> EntryDay;
+		if(EntryDay > 30 || EntryDay <= 0)
```
Same as with the IsValid stuff for games, you're not allowing the user to enter 31st as a day of the publishing date

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+				std::cin >> Choice;
+			}
+		}
+	} while (Choice = 1);
```
This is WAY too many lines to have inside a function, even more to have in just one single do-while statement.

DvdHalliday replied:
You need to separate this into like 5 different methods

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+			std::cin >> EntryYear;
+		}
+		FVideoGame NewGame(EntryGameName, EntryStudioName, EntryDay, EntryMonth, EntryYear);
+		std::cout << "You want create a Category?:" << std::endl;
```
Do you want*

DvdHalliday replied:
or rather "would you like"*

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+void FSteam::OpenCategoriesMenu() 
+{
+	system("CLS");
+	enum ChoiceOptions{CreateCategory=1,EliminateCategory};
```
Choose enum classes over simple enums, and use the prefixes

DvdHalliday replied:
also, add spaced after commas

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+	do 
+	{
+		system("CLS");
+		std::cout << "Here are Your Categories" << std::endl;
```
"your" doesn't need a capitalized Y

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+			{
+				std::cin.clear();
+				std::cin.ignore(1000,'\n');
+				std::cout << "Invalid input or title for a category, Please enter an new name for the category:" << std::endl;
```
a new name*

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+			}
+			if (CategoryContainer.GetCategoryCount() >= 5)
+			{
+				std::cout << "Sorry, You can not add more categories" << std::endl;
```
can not is not wrong but cannot should be prefered

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+			if (CategoryContainer.GetCategoryCount() >= 5)
+			{
+				std::cout << "Sorry, You can not add more categories" << std::endl;
+				std::cout << "You can Goback to the Category Menu menu pressing Enter or press 2 for go back to the main menu" << std::endl;
```
go    back. Also "press 2 TO go back"* or to get back

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+				{
+					std::cin.clear();
+					std::cin.ignore(1000, '\n');
+					std::cout << "Invalid input, enter a new choice Please:" << std::endl;
```
Please shouldn't go at the end of the sentence

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+		case EliminateCategory:
+			if(CategoryContainer.GetCategoryCount() == 0)
+			{
+				std::cout << "You dont have categories for edit " << std::endl;
```
apostrophes. Also To edit, not for edit

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+			if(CategoryContainer.GetCategoryCount() == 0)
+			{
+				std::cout << "You dont have categories for edit " << std::endl;
+				std::cout << "Press 2 for back to de Main menu" << std::endl;
```
to go back*

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+			{
+				std::cout << "You dont have categories for edit " << std::endl;
+				std::cout << "Press 2 for back to de Main menu" << std::endl;
+				std::cout << "Press enter for back to teh previous menu" << std::endl;
```
to go back*

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+				{
+					std::cout << CategoriesIndex << " " << CategoryContainer.GetCategory(CategoriesIndex).GetCategoryTitle() << std::endl;
+				}
+				std::cout << "Please select the category that you want for eliminate:" << std::endl;
```
to eliminate*

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+				}
+				CategoryContainer.EliminateCategory(SelectedToDelete);
+				std::cout << "Category Eliminated" << std::endl;
+				std::cout << "Press enter to go back to teh previous menu" << std::endl;
```
the*

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+			break;
+		}
+		
+	} while (Choice = 1);
```
Same as before, i think you could condense a lot of this into new methods

DvdHalliday commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,317 @@
+		}
+		
+	} while (Choice = 1);
+	if (Choice = 2)MenuCommand::MainMenu;
```
{ } body for the if- statement content

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.h:
```diff
@@ -0,0 +1,33 @@
+{
+	private:
+		std::string GameName;
+		std::string	StudioName;
```
Careful with the format. It's a bit inconsistent, I see that there's an indentation between the string declaration and StudioName, why is that.

A useful shortcut for formatting is `ctrl + k, ctrl + d`. This will format your whole file, because I'm seeing that the formatting is very inconsistent in this review.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.h:
```diff
@@ -0,0 +1,33 @@
+		int SetMonthCreation(const int NewMonthCration);
+		int SetYearCreation(const int NewYearCreation);
+		std::string GetFormatedDate()const;
+		bool GameIsValid()const;
```
The game is redundant. The class is already called FVideoGame, so a simple IsValid would be enough.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.cpp:
```diff
@@ -0,0 +1,75 @@
+
+FVideoGame::FVideoGame()
+{
+	GameName = "";
```
Keep in mind that classes with default constructors don't need to be explicitly initialized, as their constructors are called upon definition. So, there's really no need to explicitly initialize GameName or StudioName :)

Please fix this for the rest of your review

This is _only_ for class instances with default constructors. The way you are assigning the date fields is fine since they are not initialized by default.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.cpp:
```diff
@@ -0,0 +1,75 @@
+{
+	GameName = NewStudioName;
+}
+int FVideoGame::SetDayCreation(int NewDayCreation)
```
Why are you returning an int on these getters?

1. They don't need to return anything.
2. They aren't returning anything anyway, this is dangerous code.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.cpp:
```diff
@@ -0,0 +1,75 @@
+	if (DayCreation <= 0 || DayCreation > 30)return false;
+	else if (MonthCreation <= 0 || MonthCreation > 12)return false;
+	else if (YearCreation > 2022 || YearCreation < 1993) return false;
+	else return true;
```
Also the `else` statements are unnecessary. You are already returning, so if any of the conditions were true, none of the rest of the code would execute anyway.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.h:
```diff
@@ -0,0 +1,20 @@
+	void DebugApp();
+	void RunApp();
+	FCategory Uncategorized;
+	void DisplayCategories();
```
Why are the uncategorized instance and the DisplayCategories function public?

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.cpp:
```diff
@@ -0,0 +1,43 @@
+	else return true;
+	
+}
+void FCategoryContainer::EliminateCategory(int EliminationIndex)
```
This method is not working as intended. What if you had 4 categories and you tried to eliminate the one at index 0? Or if you tried to eliminate index 3

camodesigns committed:
Steam First Updates

Seters an getters Updated
Methods Changed
Constructors changed
Steam.h and .cc Modified

camodesigns committed:
Steam Updates

Gatters and Setters spaces and agroup

camodesigns committed:
SteamApp Review Updates

Spaces
Methods
Logics
Names

camodesigns committed:
SteamApp Adding References

Add References,
update method Delete Category

simonkytr commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Category.cpp:
```diff
@@ -0,0 +1,61 @@
+}
+FCategory::FCategory(const FCategory& OldCategory)
+{
+	GamesCount = 0;
```
here you should inizialize as `GamesCount = OldCategory.GamesCount`

JDAlzate replied:
Yup. The code is technically working correctly, but this line is completely unnecessary.

camodesigns replied:
Updated

simonkytr commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Category.cpp:
```diff
@@ -0,0 +1,61 @@
+}
+void FCategory::SetCategoryTitle(std::string NewTitle)
+{
+	Title = NewTitle;
```
this function you can set from the constructor, you don't need to create a function apart

simonkytr commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Category.cpp:
```diff
@@ -0,0 +1,61 @@
+}
+FVideoGame FCategory::GetGame(const int GameIndex) const
+{
+	return Games[GameIndex];
```
JD in the code document, explains to us about we need to change this function with references you need a const function and an Out parameter, furthermore the index

simonkytr commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.cpp:
```diff
@@ -2,13 +2,21 @@
 }
+FCategoryContainer::FCategoryContainer(FCategoryContainer& OldCategoryContainer)
+{
+	CategoriesCount = 0;
```
the same with the previous file, you should initialize as `CategoriesCount = OldCategoryContainer.CategoriesCount `

simonkytr replied:
you have to copy the memory of the array, you must use the function `memcpy_s()`

JDAlzate replied:
memcpy_s or a loop are both valid options, but yeah, you're not copying the array at all.

simonkytr commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.cpp:
```diff
@@ -25,28 +33,26 @@ bool FCategoryContainer::IsValidIndex(const int Index) const
-FCategory FCategoryContainer::GetCategory(int Index) const
+FCategory FCategoryContainer::GetCategory(const int Index) const
 {
 	return Categories[Index];
```
according to the document we have to define a function as a reference furthermore this method needs an OutParameter, you have to review the JD's document

simonkytr commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.h:
```diff
@@ -9,14 +9,16 @@ class FCategoryContainer
 
 		int GetCategoryCount() const;
-		FCategory GetCategory(int Index) const;
+		FCategory GetCategory(const int Index) const;
```
this function shouldn't be an FCategory& because this method doesn't allow make checking of the categories

simonkytr commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.h:
```diff
@@ -0,0 +1,24 @@
+	FCategory Uncategorized;
+	enum class EMenuCommand {MainMenu, AddGames, Categories ,DisplayGames, Exit};
+	EMenuCommand ActiveCommand;
+	/*enum class Action { CreateCategory = 1, DeleteCategory };
```
no need to add unnecessary comment

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.cpp:
```diff
@@ -0,0 +1,99 @@
+	ReleaseYear = PreviousVideoGame.ReleaseYear;
+
+
+}
```
Remember that empty lines and separators (look for separators in the code conventions) are necessary between methods.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.cpp:
```diff
@@ -0,0 +1,99 @@
+	{
+		return false;
+	}
+	if ((ReleaseMonth == 11 && ReleaseDay == 31) || (ReleaseMonth == 4 && ReleaseDay == 31) || (ReleaseMonth == 9 && ReleaseDay == 31) || (ReleaseMonth == 6 && ReleaseDay == 31))
```
Nice! Keep in mind that you are checking the same condition (`ReleaseDay == 31`) multiple times, plus you'll have to do all these checks if ReleaseDay is not equal to 31. You could prevent this by using an if statement and a switch statement.

Also, you missed february.
```
if (ReleaseDay == 31)
{
    switch (ReleaseMonth)
    {
        case 11:
        case 4:
        case 9:
        case 6:
        case 2:
        return false;
    }
}
```
With this method, you won't have to check anything else if you know that the ReleaseDay is not equal to 31, plus you won't have to have different checks for each potential ReleaseMonth.

camodesigns replied:
Oh I will take it in mind from now on, really thanks

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.cpp:
```diff
@@ -0,0 +1,99 @@
+	{
+		return false;
+	}
+	 if (ReleaseMonth <= 0 || ReleaseMonth > 12)
```
Small comment, but you may want to have this check before the previous check, since you don't want to check for specific months if you know that it's not a valid month.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.cpp:
```diff
@@ -0,0 +1,99 @@
+	{
+		return false;
+	}
+	else
```
This else-statement is unnecessary and a bit confusing. In all previous paths, you return false if something goes wrong, so you can simply return true here since you know that reaching the line means that nothing went wrong.
```
	if (ReleaseYear > 2022 || ReleaseYear < 1993)
	{
	    return false;
	}
	
	return true;
```

camodesigns replied:
You have the reason, updated and thanks.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.h:
```diff
@@ -0,0 +1,34 @@
+		int ReleaseYear = 0;
+	public:
+		FVideoGame();
+		FVideoGame(const FVideoGame& PreviousVideoGame);
```
I don't think that PreviousVideoGame is a very clear name. There really isn't a concept of `Previous` and `Next`. What you're doing is copying the data of one game into another. In that case maybe Other or OtherVideoGame are more appropriate names, what do you think?

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/VideoGames.h:
```diff
@@ -0,0 +1,34 @@
+		int GetReleaseYear()const;
+		std::string GetFormattedReleaseDate() const;
+
+		void SetGameName(std::string NewGameName);
```
You did a good job returning const references to strings, but remember that these should be used for parameters as well. These strings could be const references.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Category.h:
```diff
@@ -0,0 +1,28 @@
+public:
+
+	FCategory();
+	FCategory(const FCategory& OldCategory);
```
Same comment here about the name, since OldCategory is not really right either.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Category.cpp:
```diff
@@ -0,0 +1,61 @@
+	Title = OldCategory.Title;
+	GamesCount = OldCategory.GamesCount;
+}
+FCategory::FCategory(const std::string& newTitle)
```
Careful there with the use of camelCase. Remember we use PascalCase.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Category.cpp:
```diff
@@ -0,0 +1,61 @@
+	Title = newTitle;
+	GamesCount = 0;
+}
+FCategory::~FCategory() {
```
Careful there with braces.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Category.cpp:
```diff
@@ -0,0 +1,61 @@
+		GamesCount += 1;
+		return true;
+	}
+	else return false;
```
1. The else is redundant and unnecessary.
2. Braces should be used for else statements if you were to use one here.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.cpp:
```diff
@@ -0,0 +1,58 @@
+	return Categories[Index];
+}
+
+void FCategoryContainer::AddGameToCategory(const int Index, FVideoGame& VideoGame)
```
Remember that if the value of the original object shouldn't change, then you should use a const reference, rather than just a reference.

In this case, you should use `const FVideoGame& VideoGame`

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.cpp:
```diff
@@ -0,0 +1,58 @@
+
+bool FCategoryContainer::IsValidIndex(const int Index) const
+{
+	if (!std::cin.good() || Index > 4 || Index < 0) {
```
Remember the braces conventions

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.cpp:
```diff
@@ -0,0 +1,58 @@
+	if (!std::cin.good() || Index > 4 || Index < 0) {
+		return false;
+	}
+	else
```
This else is also redundant, could just be the return alone :)

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/CategoryContainer.cpp:
```diff
@@ -0,0 +1,58 @@
+
+bool FCategoryContainer::DeleteCategory(const int EliminationIndex)
+{
+	if (EliminationIndex < CategoriesCount)
```
What if I use -1 as an index?

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Main.cpp:
```diff
@@ -0,0 +1,8 @@
+{
+	FSteam Steam;
+	Steam.RunApp();
+	//Steam.Actions();
```
Please remove unused, commented code. It pollutes the codebase 👀

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.h:
```diff
@@ -0,0 +1,24 @@
+	Action ActionChoice;*/
+public:	
+	FSteam();
+	FSteam(FSteam& oldSteam);
```
Keep in mind that copy constructors should use const references, not just references. You don't want to modify the other object

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,408 @@
+		std::cout << "Please enter the name of the Studio " << std::endl;
+		std::string EntryStudioName;
+		std::getline(std::cin, EntryStudioName);
+		while (!std::cin.good() || EntryGameName=="")
```
This validation code is repeated many many many times. This is when a function should be considered. You want to avoid code duplication as much as possible (which should be ideally always).

JDAlzate replied:
This is also an incredibly long method. This would be simplified a whole lot if you had a function for validation.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,408 @@
+
+			}
+		}
+		else if (ChoiceAction == 3) {
```
Braces

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,408 @@
+
+
+	} while (Choice == 1);
+	if (Choice == 2)ActiveCommand = EMenuCommand::MainMenu;
```
Braces

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,408 @@
+	if (Choice == 2)ActiveCommand = EMenuCommand::MainMenu;
+	
+}
+
```
There's extra whitespace here. One space should be enough + remember to use separators next time.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,408 @@
+			}
+			else
+			{
+				for (int CategoriesIndex = 0; CategoriesIndex < 5; CategoriesIndex += 1)
```
You're hardcoding the max index for your categories, even though you have a variable for that in the category container. If your category count changed in the category container, your code would have a bug here.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,408 @@
+
+void FSteam::OpenVisualiceGamesMenu()
+{
+		for (int CategoryIndex = 0; CategoryIndex < 5; CategoryIndex+=1)
```
Same thing about hardcoding your max index.

JDAlzate commented on Month2/week1/Month2-Project Week 1-Remembering the Bases/VideoGame/Steam.cpp:
```diff
@@ -0,0 +1,408 @@
+{
+		for (int CategoryIndex = 0; CategoryIndex < 5; CategoryIndex+=1)
+		{
+			if (CategoryContainer.GetCategoryCount() != 0)
```
Remind me to take a look at this part with you tomorrow please :) It's working fine, but there's a few better ways to achieve this.

camodesigns replied:
Updated

JDAlzate reviewed (changes requested):
I see a lot of improvement compared to last week :) Still some things to change, but we're on track here 👍

camodesigns committed:
Code review Updates

Spaces
Names
Structures
Methods
Returns
Brase
Create methods
Eliminate unnecesary lynes

camodesigns committed:
GetGameAt(const int Index, FVideoGame& OutGame) added

Method created
